### DISCUSSION TITLE
Is there a way to return a function?

### DISCUSSION THREAD

**Discussion by jjangga0214** [napi-rs/napi-rs#1646] [Q&A]
Hi!

I'd like to declare a function whose return type is another function.

But the compiler complains.

<img width="823" alt="스크린샷 2023-07-08 오후 7 23 45" src="https://github.com/napi-rs/napi-rs/assets/28584151/acc7d572-3b7b-4650-b501-cbc204dff3f2">

Is there a proper valid way?

One workaround is this. 

```rs 
#[napi]
pub struct Foo;

#[napi]
impl Foo {
  #[napi]
  pub fn foo(&self) {
    println!("foo");
  }
}

#[napi]
pub fn foo() -> Foo {
  Foo {}
}
```

```js
import { foo,bar } from './index.js'

const res = foo()
res.foo()
```

But this requires consumer to call `res.foo()`, not `res()`. 

To solve this issue, this solution may be used.

```rs 
#![feature(unboxed_closures)]
#![feature(fn_traits)]

#[napi]
pub struct Foo;

#[napi]
impl Fn for Foo {
  #[napi]
  pub fn call(&self, _args: ()) {
    println!("foo");
  }
}

#[napi]
pub fn foo() -> Foo {
  Foo {}
}
```

```js
import { foo } from './index.js'

const res = foo()
res()
```

But this requires `#![feature]` flag, which means depending on unstable version of Rust.

And I feel the first and second solutions are all boilerplate.

I just want to simply write this.

```rs
#[napi]
pub fn foo() {
  Box::new(|| println!("foo"))
}
```

Is there a way?

Thanks :)

**Comment by Brooooooklyn** [comment]
See this for example: https://github.com/napi-rs/napi-rs/blob/main/examples/napi/src/callback.rs#L51

**Reply by jjangga0214** [reply]
Thank you! 
However, is there a way to return lambda?
